documentation/Using-the-SDK/Advanced/Creating-Your-Own-DataSource.md

```diff
@@ -0,0 +1,204 @@
+
+````cs
+// we are decorating Custom Data Source classes, so we must have the correct
+// attribute targets
+[AttributeUsage(AttributeTargets.Class)]
+public sealed class WebServiceDataSourceAttribute
+    : DataSourceAttribute
+{
+    // our constructor will pass the DataSource Type to the base class. We
+    // can take any other parameters that we like to help denote our data. In
+    // this case, we simply say that data on the given server is supported.
+    public WebServiceDataSourceAttribute(
+        string serviceUri)
+        : base(typeof(WebServiceDataSource))
+    {
+        this.ServiceUri = new Uri(serviceUri);
+    }
+
+    public Uri ServiceUri { get; }
+
+    // implement this method to eagerly filter out data sources that are not
+    // accepted by the decorated Custom Data Source.
+    public override bool Accepts(IDataSource dataSource)
+    {
+        // IDataSource will always be of the Type based to the base
+        // class in our constructor.
+        var web = dataSource as WebServiceDataSource;
+
+        // This check is for defensive programming in case dataSource is
+        // ever null.
+        if (web is null)
+        {
+            return false;
+        }
+
+        // We only accept WebServiceDataSources that target our declared service.
+        return this.ServiceUri == web.ServiceUri;
+    } 
+}
+````
+
+# Using Your Data Source 
+
+Now that we have our new Data Source, we will see how to use our new Data Source
+with the Engine.
+
+````cs
+
+namespace Sample
+{
+    [CustomDataSource(
+        /* Guid here */,
+        /* Name here */,
+        /* Description here */)]
+    [WebServiceDataSource("https://www.contoso.com")]
+    public class ContosoDataSource
+        : CustomDataSourceBase
+    {
+        // implementation elided
+    }
+
+    public class Program
+    {
+        public static void Main(string[] args)
+        {
+            //
+            // Prep our data sources. These could come from the
+            // command line, or be hard coded.
+            //
+
+            var data1 = new WebDataSource("https://www.microsoft.com");
+            var data2 = new WebDataSource("https://www.contoso.com");
+
+            //
+            // Create the Engine.
+            //
+            var engine = Engine.Create();
+
+            //
+            // Add our data
+            //
+
+            // data1 will not get associated to ContosoDataSource...
+            engine.AddDataSource(data1);
+
+            // ...but data2 will.
+            engine.AddDataSource(data2);
+
+            //
+            // add any cookers, etc.
+            //
+
+            //
+            // Process our data
+            //
+            var result = engine.Process();
+
+            //
+            // Deal with the processed data.
+            //
+        }
+    }
+}
+
+````
+
+# Conclusion
+
+We have seen how to create a new kind of Data Source an how to seamlessly
```

Comment:
after its made, maybe link back to advanced/overview.md

Reply:
Will do.